src/gtest/test_checktransaction.cpp

```diff
@@ -45,9 +43,309 @@ class MockCValidationState : public CValidationState {
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.vpour[0].serials.at(0) = uint256S("0000000000000000000000000000000000000000000000000000000000000000");
+    mtx.vpour[1].serials.at(0) = uint256S("0000000000000000000000000000000000000000000000000000000000000000");
+
+    CTransaction tx(mtx);
+
+    MockCValidationState state;
+    EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-pours-serials-duplicate", false)).Times(1);
+    CheckTransactionWithoutProofVerification(tx, state);
+}
+
+TEST(checktransaction_tests, bad_cb_has_pours) {
+    CMutableTransaction mtx = GetValidTransaction();
+    // Make it a coinbase.
+    mtx.vin.resize(1);
+    mtx.vin[0].prevout.SetNull();
+
+    mtx.vpour.resize(1);
+
+    CTransaction tx(mtx);
+    EXPECT_TRUE(tx.IsCoinBase());
+
+    MockCValidationState state;
+    EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-cb-has-pours", false)).Times(1);
+    CheckTransactionWithoutProofVerification(tx, state);
+}
+
+TEST(checktransaction_tests, bad_cb_empty_scriptsig) {
+    CMutableTransaction mtx = GetValidTransaction();
+    // Make it a coinbase.
+    mtx.vin.resize(1);
+    mtx.vin[0].prevout.SetNull();
+
+    mtx.vpour.resize(0);
+
+    CTransaction tx(mtx);
+    EXPECT_TRUE(tx.IsCoinBase());
+
+    MockCValidationState state;
+    EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-cb-length", false)).Times(1);
+    CheckTransactionWithoutProofVerification(tx, state);
+}
+
+TEST(checktransaction_tests, bad_txns_prevout_null) {
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.vin[1].prevout.SetNull();
+
+    CTransaction tx(mtx);
+    EXPECT_FALSE(tx.IsCoinBase());
+
+    MockCValidationState state;
+    EXPECT_CALL(state, DoS(10, false, REJECT_INVALID, "bad-txns-prevout-null", false)).Times(1);
+    CheckTransactionWithoutProofVerification(tx, state);
+}
+
+TEST(checktransaction_tests, bad_txns_invalid_joinsplit_signature) {
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.joinSplitSig[0] += 1;
+    CTransaction tx(mtx);
+
+    MockCValidationState state;
+    EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "bad-txns-invalid-joinsplit-signature", false)).Times(1);
+    CheckTransactionWithoutProofVerification(tx, state);
+}
+
+TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    CMutableTransaction mtx = GetValidTransaction();
+
+    // Copied from libsodium/crypto_sign/ed25519/ref10/open.c
+    static const unsigned char L[32] =
+      { 0xed, 0xd3, 0xf5, 0x5c, 0x1a, 0x63, 0x12, 0x58,
+        0xd6, 0x9c, 0xf7, 0xa2, 0xde, 0xf9, 0xde, 0x14,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10 };
+
+    // Add L to S, which starts at mtx.joinSplitSig[32].
+    unsigned int s = 0;
+    for (size_t i = 0; i < 32; i++) {
+        s = mtx.joinSplitSig[32 + i] + L[i] + (s >> 8);
+        mtx.joinSplitSig[32 + i] = s & 0xff;
+    }
```

Comment:
I could add an `EXPECT_TRUE( s >> 8 == 0 );` here, but it's guaranteed because the original S should be less than L and 2*L < 2^256.